src/items/associated-items.md

```diff
@@ -0,0 +1,260 @@
+usage in the rest of the function declaration. The generics, parameter list,
+return type, and where clause must be the same in the associated function
+definition.
+
+An *associated function definiton* is written as an associated function
+declaration, but instead of a `;`, there is a [block] that evaluates to the
+return type.
+
+An example of a common associated function is the `new` function that returns
+a value of the type the associated function is associated with.
+
+```rust
+struct Struct {
+    field: i32;
+}
+
+impl Struct {
+    fn new() -> Struct {
+        Struct {
+            field: 0i32
+        }
+    }
+}
+```
+
+When the associated function is declared on a trait, the function can be called
+on the trait. When this happens, it is substituted for
+`<_ as Trait>::function_name`. 
+
+```rust
+trait Num {
+    fn from_i32(n: i32) -> Self;
+}
+impl Num for f64 {
+    fn from_i32(n: i32) -> f64 { n as f64 }
+}
+let x: f64 = Num::from_i32(42);
+let x: f64 = f64::from_i32(42);
+```
+
+Associated functions whose first parameter is named `self` are called *methods*
+and may be invoked using the [method call operator], for example, `x.foo()`, as
+well as the usual function call notation.
+
+When the first parameter is named `self`, the following shorthands may be used:
+
+* `self` -> `self: Self`
+* `&self` -> `self: &Self`
+* `&mut self` -> `&mut Self`
+* `Box<self>` -> `self: Box<Self>`
+
+Consider the following trait:
+
+```rust
+# type Surface = i32;
+# type BoundingBox = i32;
+trait Shape {
+    fn draw(&self, Surface);
+    fn bounding_box(&self) -> BoundingBox;
+}
+```
+
+This defines a trait with two methods. All values that have [implementations]
+of this trait in scope can have their `draw` and `bounding_box` methods called.
+
+```rust
+# type Surface = i32;
+# type BoundingBox = i32;
+# trait Shape {
+#     fn draw(&self, Surface);
+#     fn bounding_box(&self) -> BoundingBox;
+# }
+
+struct Circle {
+    // ...
+}
+
+impl Shape for Circle {
+    // ...
+#   fn draw(&self, Surface) -> {}
+#   fn bounding_box(&self) -> BoundingBox { 0i32; }
+}
+
+# impl Box {
+#     fn new() -> Circle { Circle{} }
+}
+
+let circle_shape = Circle::new();
+let bounding_box = circle_shape.bounding_box();
+```
+
+## Associated Types
+
+*Associated types* are [type aliases] associated with another type. Associated
+types cannot be defined in [inherent implementations] nor can they be given a
+default implementation in traits.
+
+An *associated type declaration* is written as `type`, then an [identifier], and
+finally an optional trait bounds.
+
+The identifier is the name of the declared type alias. The optional trait bounds
+must be fulfilled by the implementations of the type alias.
+
+An *associated type definition* is written as `type`, then an [identifier], then
+an `=`, and finally a [type].
+
+If an item `Item` has an associated type `Assoc`, then `Item::Assoc` is a type
```

Comment:
Does `Self` count as a built-in associated type? I feel like it should be mentioned here.